### ISSUE TITLE
--include-easyblocks broken with recent setuptools (19.4 or newer)

### ISSUE THREAD

**Issue by ocaisa** [hpcugent/easybuild-framework#1678]
With the latest code in develop, `--include-easyblocks` becomes broken when I load GC3Pie (which includes a module for setuptools 20.3, CentOS 7 setuptools is 0.9.8):
```
-bash-4.2$ eb --search GC3
== temporary log file in case of crash /tmp/eb-RT6Y3q/easybuild-B3kHwG.log
== Searching (case-insensitive) for 'GC3' in /work/zam/swmanage/EasyBuild/Golden_Repo/Stage1 
== Searching (case-insensitive) for 'GC3' in /usr/local/software/jureca/Stage1/eb_repo 
== Searching (case-insensitive) for 'GC3' in /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs 
 * /work/zam/swmanage/EasyBuild/Golden_Repo/Stage1/GC3Pie-2.4.2.eb
 * /usr/local/software/jureca/Stage1/eb_repo/GC3Pie/GC3Pie-2.4.2.eb
 * /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs/g/GC3Pie/GC3Pie-2.2.3.eb
 * /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs/g/GC3Pie/GC3Pie-2.3.eb
 * /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs/g/GC3Pie/GC3Pie-2.4.0.eb
 * /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs/g/GC3Pie/GC3Pie-2.4.1.eb
 * /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs/g/GC3Pie/GC3Pie-2.4.2.eb
== Temporary log file(s) /tmp/eb-RT6Y3q/easybuild-B3kHwG.log* have been removed.
== Temporary directory /tmp/eb-RT6Y3q has been removed.
-bash-4.2$ module load GC3Pie
-bash-4.2$ eb --search GC3
ERROR: Failed to parse configuration options: '/homeb/zam/swmanage/software/easybuild-easyblocks/easybuild' is not in list

```

**Comment by ocaisa** [comment]
May be connected to #1677

**Comment by ocaisa** [comment]
pdb output is
```
(Pdb) print sys.path
['', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib64/python2.7/site-packages/SQLAlchemy-1.0.8-py2.7-linux-x86_64.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib64/python2.7/site-packages/simplejson-3.8.0-py2.7-linux-x86_64.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/setuptools-20.3-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/ecdsa-0.13-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/paramiko-1.16.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/prettytable-0.7.2-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/pyCLI-devel-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/parsedatetime-1.5-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/boto-2.38.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/pip-7.1.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/Babel-2.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/pytz-2016.1-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/requests-2.7.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/iso8601-0.1.10-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/argparse-1.3.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/six-1.9.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/gc3pie-2.4.2-py2.7.egg', '/homeb/zam/swmanage/software/vsc-base/lib', '/homeb/zam/swmanage/software/easybuild-easyconfigs', '/homeb/zam/swmanage/software/easybuild-easyblocks', '/homeb/zam/swmanage/software/easybuild-framework', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib64/python2.7/site-packages', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages', '/usr/lib64/python27.zip', '/usr/lib64/python2.7', '/usr/lib64/python2.7/plat-linux2', '/usr/lib64/python2.7/lib-tk', '/usr/lib64/python2.7/lib-old', '/usr/lib64/python2.7/lib-dynload', '/usr/lib64/python2.7/site-packages', '/usr/lib/python2.7/site-packages']
(Pdb) n
> /homeb/zam/swmanage/software/easybuild-framework/easybuild/tools/include.py(177)include_easyblocks()
-> fixup_namespace_packages(easyblocks_path)
(Pdb) n
ValueError: "'/homeb/zam/swmanage/software/easybuild-easyblocks/easybuild' is not in list"
> /homeb/zam/swmanage/software/easybuild-framework/easybuild/tools/include.py(177)include_easyblocks()
-> fixup_namespace_packages(easyblocks_path)
```

**Comment by ocaisa** [comment]
Without GC3Pie loaded, the pdb output is:
```
-bash-4.2$ eb --search GC3
> /homeb/zam/swmanage/software/easybuild-framework/easybuild/tools/include.py(176)include_easyblocks()
-> sys.path.insert(0, easyblocks_path)
(Pdb) print sys.path
['', '/homeb/zam/swmanage/software/vsc-base/lib', '/homeb/zam/swmanage/software/easybuild-easyconfigs', '/homeb/zam/swmanage/software/easybuild-easyblocks', '/homeb/zam/swmanage/software/easybuild-framework', '/usr/lib64/python27.zip', '/usr/lib64/python2.7', '/usr/lib64/python2.7/plat-linux2', '/usr/lib64/python2.7/lib-tk', '/usr/lib64/python2.7/lib-old', '/usr/lib64/python2.7/lib-dynload', '/usr/lib64/python2.7/site-packages', '/usr/lib/python2.7/site-packages']
(Pdb) cont
== temporary log file in case of crash /tmp/eb-YELly7/easybuild-6xoxCI.log
```

**Comment by ocaisa** [comment]
After the fix coming for #1677 pdb output is:
```
-bash-4.2$ eb --search GC3
> /homeb/zam/swmanage/software/easybuild-framework/easybuild/tools/include.py(176)include_easyblocks()
-> sys.path.insert(0, easyblocks_path)
(Pdb) print sys.path
['', '/homeb/zam/swmanage/software/vsc-base/lib', '/homeb/zam/swmanage/software/easybuild-easyconfigs', '/homeb/zam/swmanage/software/easybuild-easyblocks', '/homeb/zam/swmanage/software/easybuild-framework', '/usr/lib64/python27.zip', '/usr/lib64/python2.7', '/usr/lib64/python2.7/plat-linux2', '/usr/lib64/python2.7/lib-tk', '/usr/lib64/python2.7/lib-old', '/usr/lib64/python2.7/lib-dynload', '/usr/lib64/python2.7/site-packages', '/usr/lib/python2.7/site-packages']
(Pdb) cont
== temporary log file in case of crash /tmp/eb-79Lv9A/easybuild-m8thmL.log
== Searching (case-insensitive) for 'GC3' in /work/zam/swmanage/EasyBuild/Golden_Repo/Stage1 
== Searching (case-insensitive) for 'GC3' in /usr/local/software/jureca/Stage1/eb_repo 
== Searching (case-insensitive) for 'GC3' in /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs 
 * /work/zam/swmanage/EasyBuild/Golden_Repo/Stage1/GC3Pie-2.4.2.eb
 * /usr/local/software/jureca/Stage1/eb_repo/GC3Pie/GC3Pie-2.4.2.eb
 * /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs/g/GC3Pie/GC3Pie-2.2.3.eb
 * /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs/g/GC3Pie/GC3Pie-2.3.eb
 * /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs/g/GC3Pie/GC3Pie-2.4.0.eb
 * /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs/g/GC3Pie/GC3Pie-2.4.1.eb
 * /homeb/zam/swmanage/software/easybuild-easyconfigs/easybuild/easyconfigs/g/GC3Pie/GC3Pie-2.4.2.eb
== Temporary log file(s) /tmp/eb-79Lv9A/easybuild-m8thmL.log* have been removed.
== Temporary directory /tmp/eb-79Lv9A has been removed.
-bash-4.2$ module load GC3Pie
-bash-4.2$ eb --search GC3
> /homeb/zam/swmanage/software/easybuild-framework/easybuild/tools/include.py(176)include_easyblocks()
-> sys.path.insert(0, easyblocks_path)
(Pdb) print sys.path
['', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib64/python2.7/site-packages/SQLAlchemy-1.0.8-py2.7-linux-x86_64.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib64/python2.7/site-packages/simplejson-3.8.0-py2.7-linux-x86_64.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/setuptools-20.3-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/ecdsa-0.13-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/paramiko-1.16.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/prettytable-0.7.2-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/pyCLI-devel-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/parsedatetime-1.5-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/boto-2.38.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/pip-7.1.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/Babel-2.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/pytz-2016.1-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/requests-2.7.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/iso8601-0.1.10-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/argparse-1.3.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/six-1.9.0-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages/gc3pie-2.4.2-py2.7.egg', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib64/python2.7/site-packages', '/usr/local/software/jureca/Stage1/software/GC3Pie/2.4.2/lib/python2.7/site-packages', '/homeb/zam/swmanage/software/vsc-base/lib', '/homeb/zam/swmanage/software/easybuild-easyconfigs', '/homeb/zam/swmanage/software/easybuild-easyblocks', '/homeb/zam/swmanage/software/easybuild-framework', '/usr/lib64/python27.zip', '/usr/lib64/python2.7', '/usr/lib64/python2.7/plat-linux2', '/usr/lib64/python2.7/lib-tk', '/usr/lib64/python2.7/lib-old', '/usr/lib64/python2.7/lib-dynload', '/usr/lib64/python2.7/site-packages', '/usr/lib/python2.7/site-packages']
(Pdb) cont
ERROR: Failed to parse configuration options: '/homeb/zam/swmanage/software/easybuild-easyblocks/easybuild' is not in list
```

**Comment by ocaisa** [comment]
This only seems to occur for setuptools 20.3, just tested 18.1 and there was no problem

**Comment by ocaisa** [comment]
The _error_ is introduced in 19.4, but we may actually be relying on a bug. Likely commit/bug is at https://bitbucket.org/pypa/setuptools/issues/207/141-introduces-regression-in-importing

**Comment by boegel** [comment]
@ocaisa are you using the latest `develop` of the easyblocks repo too here? In particular, do you have `declare_namespace` in `easybuild/__init__.py` and `easybuild/easyblocks/__init__.py`?

**Comment by boegel** [comment]
I'm able to reproduce this issue with the fix from #1679 included, provided I define `$EASYBUILD_INCLUDE_EASYBLOCKS` ...

What's even more worrying is that *all* of the framework unit tests fail (regardless of having `$EASYBUILD_INCLUDE_EASYBLOCKS` defined or not), if setuptools 19.4 or higher is used.

Can you confirm the latter @ocaisa?

It looks like we have another blocker on our hands... :(

**Comment by boegel** [comment]
Full traceback:

```
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/private/tmp/1677/software/EasyBuild/2.7.0.dev0/lib/python2.7/site-packages/easybuild_framework-2.7.0.dev0-py2.7.egg/easybuild/main.py", line 408, in <module>
    main()
  File "/private/tmp/1677/software/EasyBuild/2.7.0.dev0/lib/python2.7/site-packages/easybuild_framework-2.7.0.dev0-py2.7.egg/easybuild/main.py", line 197, in main
    eb_go = eboptions.parse_options(args=args)
  File "/private/tmp/1677/software/EasyBuild/2.7.0.dev0/lib/python2.7/site-packages/easybuild_framework-2.7.0.dev0-py2.7.egg/easybuild/tools/options.py", line 981, in parse_options
    go_args=args, error_env_options=True, error_env_option_method=raise_easybuilderror)
  File "/private/tmp/1677/software/EasyBuild/2.7.0.dev0/lib/python2.7/site-packages/easybuild_framework-2.7.0.dev0-py2.7.egg/easybuild/tools/options.py", line 178, in __init__
    super(EasyBuildOptions, self).__init__(*args, **kwargs)
  File "/private/tmp/1677/software/EasyBuild/2.7.0.dev0/lib/python2.7/site-packages/vsc_base-2.4.18-py2.7.egg/vsc/utils/generaloption.py", line 985, in __init__
    self.postprocess()
  File "/private/tmp/1677/software/EasyBuild/2.7.0.dev0/lib/python2.7/site-packages/easybuild_framework-2.7.0.dev0-py2.7.egg/easybuild/tools/options.py", line 588, in postprocess
    self._postprocess_include()
  File "/private/tmp/1677/software/EasyBuild/2.7.0.dev0/lib/python2.7/site-packages/easybuild_framework-2.7.0.dev0-py2.7.egg/easybuild/tools/options.py", line 639, in _postprocess_include
    include_easyblocks(self.tmpdir, self.options.include_easyblocks)
  File "/private/tmp/1677/software/EasyBuild/2.7.0.dev0/lib/python2.7/site-packages/easybuild_framework-2.7.0.dev0-py2.7.egg/easybuild/tools/include.py", line 177, in include_easyblocks
    fixup_namespace_packages(easyblocks_path)
  File "build/bdist.macosx-10.10-intel/egg/pkg_resources/__init__.py", line 2237, in fixup_namespace_packages
  File "build/bdist.macosx-10.10-intel/egg/pkg_resources/__init__.py", line 2235, in fixup_namespace_packages
  File "build/bdist.macosx-10.10-intel/egg/pkg_resources/__init__.py", line 2194, in _handle_ns
  File "build/bdist.macosx-10.10-intel/egg/pkg_resources/__init__.py", line 2192, in sort_key
ValueError: '/private/tmp/1677/software/EasyBuild/2.7.0.dev0/lib/python2.7/site-packages/easybuild_easyblocks-2.7.0.dev0-py2.7.egg/easybuild' is not in list
```

**Comment by boegel** [comment]
I reported this problem at https://bitbucket.org/pypa/setuptools/pull-requests/167/possible-fix-for-207/activity#comment-15979332

**Comment by ocaisa** [comment]
I pulled everything before I started playing with it today

**Comment by chrisacbr** [comment]
FWIW, I think this is a result of the `easybuild.easyblocks` module adding additional subdirectories to its package search path:
```python
subdirs = [chr(l) for l in range(ord('a'), ord('z') + 1)] + ['0']
for subdir in subdirs:
    __path__ = extend_path(__path__, '%s.%s' % (__name__, subdir))
```
Then, when attempting to sort the search path into the same order as `sys.path`, the setuptools code assumes that the paths match exactly the module hierarchy (ie. no additional subdirectories):
```console
> /local/z00/easybuild/software/EasyBuild/2.7.0.dev0/lib/python2.6/site-packages/setuptools-20.3-py2.6.egg/pkg_resources/__init__.py(2036)position_in_sys_path()
-> parts = parts[:-(package_name.count('.') + 1)]
(Pdb) p package_name
'easybuild.easyblocks'
(Pdb) p parts[-2:]
['easybuild', 'easyblocks']
(Pdb) n
> /local/z00/easybuild/software/EasyBuild/2.7.0.dev0/lib/python2.6/site-packages/setuptools-20.3-py2.6.egg/pkg_resources/__init__.py(2037)position_in_sys_path()
-> return sys_path.index(_normalize_cached(os.sep.join(parts)))
(Pdb) p os.sep.join(parts) in sys_path
True
(Pdb) c
> /local/z00/easybuild/software/EasyBuild/2.7.0.dev0/lib/python2.6/site-packages/setuptools-20.3-py2.6.egg/pkg_resources/__init__.py(2036)position_in_sys_path()
-> parts = parts[:-(package_name.count('.') + 1)]
(Pdb) p parts[-2:]
['easyblocks', 'a']
(Pdb) n
> /local/z00/easybuild/software/EasyBuild/2.7.0.dev0/lib/python2.6/site-packages/setuptools-20.3-py2.6.egg/pkg_resources/__init__.py(2037)position_in_sys_path()
-> return sys_path.index(_normalize_cached(os.sep.join(parts)))
(Pdb) p os.sep.join(parts) in sys_path
False
(Pdb) p os.sep.join(parts)
'/local/z00/easybuild/software/EasyBuild/2.7.0.dev0/lib/python2.6/site-packages/easybuild_easyblocks-2.7.0.dev0-py2.6.egg/easybuild'
(Pdb) n
ValueError: 'list.index(x): x not in list'
```

**Comment by boegel** [comment]
Thanks @chrisacbr, that's really helpful.

The setuptools code is using the 'depth' of the package name, which is exactly what we're fiddling with (we're removing the letter subdirectories).

That does mean there's an easy (yet kind of dirty) fix for this problem...
We just inject the `/path/to/.egg/easybuild` directory in `sys.path` in `easybuild/easyblocks/__init__.py`.

That works, I just tried.

**Comment by boegel** [comment]
More specifically, we can apply this patch to framework:

```diff
diff --git a/easybuild/tools/include.py b/easybuild/tools/include.py
index 688318f..6f5ed66 100644
--- a/easybuild/tools/include.py
+++ b/easybuild/tools/include.py
@@ -65,13 +65,16 @@ pkg_resources.declare_namespace(__name__)
 # this is required because of the way in which the easyblock Python modules are organised in the easybuild-easyblocks
 # repository, i.e. in first-letter subdirectories
 EASYBLOCKS_PKG_INIT_BODY = """
+import os
 import pkg_resources
 import pkgutil
+import sys
 
 # extend path so Python finds our easyblocks in the subdirectories where they are located
 subdirs = [chr(l) for l in range(ord('a'), ord('z') + 1)] + ['0']
 for subdir in subdirs:
     __path__ = pkgutil.extend_path(__path__, '%s.%s' % (__name__, subdir))
+sys.path.append(os.path.dirname(os.path.dirname(__file__)))
 
 del l, subdir, subdirs
 
```

and this to easyblocks:

```diff
diff --git a/easybuild/easyblocks/__init__.py b/easybuild/easyblocks/__init__.py
index 7380d65..0ea1b24 100644
--- a/easybuild/easyblocks/__init__.py
+++ b/easybuild/easyblocks/__init__.py
@@ -32,6 +32,7 @@ easybuild.easyblocks package declaration
 @author: Jens Timmerman (Ghent University)
 """
 import os
+import sys
 import pkg_resources
 from distutils.version import LooseVersion
 from pkgutil import extend_path
@@ -77,6 +78,7 @@ else:
 subdirs = [chr(l) for l in range(ord('a'), ord('z') + 1)] + ['0']
 for subdir in subdirs:
     __path__ = extend_path(__path__, '%s.%s' % (__name__, subdir))
+sys.path.append(os.path.dirname(os.path.dirname(__file__)))
 
 del subdir, subdirs, l, git_rev
 

```

It's dirty, but harmless; side-effect is that import statements like `import easyblocks...` will also sometimes work, but we don't use those at all (nor should we).
We can even do it conditionally on the `setuptools` version...

Thoughts @chrisacbr, @ocaisa, @wpoely86?

**Comment by ocaisa** [comment]
If it removes the need to restrict the setuptools version then I prefer it, that's a much less likely hole for a normal user to fall in to.

**Comment by boegel** [comment]
@ocaisa Can you verify that it fixes your problem?

**Comment by ocaisa** [comment]
Yes, problem is fixed

**Comment by boegel** [comment]
I opened a bug report for this problem at https://bitbucket.org/pypa/setuptools/issues/520/fixup_namespace_packages-may-trigger

**Comment by boegel** [comment]
While playing around with this, I noticed that the `fixup_namespace_packages` doesn't seem to be needed at all...

So, I dropped it, and only included the hackish workaround for the setuptools bug in the tests (where `fixup_namespace_packages` *is* required), cfr. #1680.

**Comment by boegel** [comment]
fixed in #1680, thanks for reporting this @ocaisa, and @chrisacbr for helping out!
